infrastructure/src/main/java/org/corfudb/infrastructure/LogReplicationServer.java

```diff
@@ -0,0 +1,114 @@
+
+        // TODO (hack): where can we obtain the local corfu endpoint? site manager? or can we always assume port is 9000?
+        String corfuPort = serverContext.getLocalEndpoint().equals("localhost:9020") ? ":9001" : ":9000";
+        String corfuEndpoint = serverContext.getNodeLocator().getHost() + corfuPort;
+        log.info("Initialize Sink Manager with CorfuRuntime to {}", corfuEndpoint);
+        this.sinkManager = new LogReplicationSinkManager(corfuEndpoint, config);
+    }
+
+    private LogReplicationConfig getDefaultLogReplicationConfig() {
+        Set<String> streamsToReplicate = new HashSet<>(Arrays.asList("Table001", "Table002", "Table003"));
+        return new LogReplicationConfig(streamsToReplicate, UUID.randomUUID(), UUID.randomUUID());
+    }
+
+    /* ************ Override Methods ************ */
+
+    @Override
+    public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
+        return getState() == ServerState.READY;
+    }
+
+    @Override
+    protected void processRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        executor.submit(() -> getHandler().handle(msg, ctx, r));
+    }
+
+    @Override
+    public void shutdown() {
+        super.shutdown();
+        executor.shutdown();
+    }
+
+    /* ************ Server Handlers ************ */
+
+    @ServerHandler(type = CorfuMsgType.LOG_REPLICATION_ENTRY)
+    private void handleLogReplicationEntry(CorfuPayloadMsg<LogReplicationEntry> msg, ChannelHandlerContext ctx, IServerRouter r) {
+        log.info("Log Replication Entry received by Server.");
+
+        LogReplicationEntry ack = sinkManager.receive(msg.getPayload());
+
+        if (ack != null) {
+            long ts = ack.getMetadata().getMessageMetadataType().equals(MessageType.LOG_ENTRY_REPLICATED) ?
+                    ack.getMetadata().getTimestamp() : ack.getMetadata().getSnapshotTimestamp();
+            log.info("Sending ACK {} on {} to Client ", ack.getMetadata(), ts);
+            r.sendResponse(ctx, msg, CorfuMsgType.LOG_REPLICATION_ENTRY.payloadMsg(ack));
+        }
+    }
+
+    @ServerHandler(type = CorfuMsgType.LOG_REPLICATION_NEGOTIATION_REQUEST)
+    private void handleLogReplicationNegotiationRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        log.info("Log Replication Negotiation Request received by Server.");
+        LogReplicationNegotiationResponse response = new LogReplicationNegotiationResponse(0, 0);
+        r.sendResponse(ctx, msg, CorfuMsgType.LOG_REPLICATION_NEGOTIATION_RESPONSE.payloadMsg(response));
+    }
+
+    @ServerHandler(type = CorfuMsgType.LOG_REPLICATION_QUERY_LEADERSHIP)
+    private void handleLogReplicationQueryLeadership(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        log.info("******Log Replication Query Leadership Request received by Server.");
+        LogReplicationQueryLeaderShipResponse resp = new LogReplicationQueryLeaderShipResponse(0, getSinkManager().isLeader());
```

Comment:
to fix the epoch value.  Will the remote sink manager query every server node until it finds the leader?

Also, why do we only consider sink leader?  Cant source and sink leaders be different?  Which leader is being requested here?